cores/esp8266/core_esp8266_postmortem.cpp

```diff
@@ -94,16 +99,22 @@ void __wrap_system_restart_local() {
            break into GDB here. */
         raise_exception();
     }
+#endif
 
     struct rst_info rst_info;
     memset(&rst_info, 0, sizeof(rst_info));
-    system_rtc_mem_read(0, &rst_info, sizeof(rst_info));
-    if (rst_info.reason != REASON_SOFT_WDT_RST &&
-        rst_info.reason != REASON_EXCEPTION_RST &&
-        rst_info.reason != REASON_WDT_RST)
+    if (fake_rst_reason == FAKE_REASON_NONE)
     {
-        return;
+        system_rtc_mem_read(0, &rst_info, sizeof(rst_info));
+        if (rst_info.reason != REASON_SOFT_WDT_RST &&
+            rst_info.reason != REASON_EXCEPTION_RST &&
+            rst_info.reason != REASON_WDT_RST)
+        {
+            return;
```

Comment:
Since you are in this code I thought I would raise this question `return`.

I have been through this module a few times and I don't understand how a `return` would work from `system_restart_local()`? Is there anything to return to? 

I searched for quite a while for info on `system_restart_local` but I could never find anything official. However, I did find an old pseudo C interpretation of disassembled ~v1 SDK. After taking a second look at the pseudo C code, `system_restart_local()` appears to be called from either a scheduled timer callback 100ms after `system_restart` is called or from Soft WDT or some system startup failure. I would expect more paths but I don't see them yet. That's all the info I could find and it is probably outdated for what we have today.

Here is an outline of the logic flow I am reading:

`system_restart()` (FLASH)
  * shutdown wifi interface
  * shutdown wifi timer callbacks
  * Start a callback timer to run `system_restart_local()` once, after 100ms delay.
  * returns to caller
    * SDK notes: "... do not call other functions after calling this API"

`system_restart_local()` (FLASH) - Runs from a timer callback or called directly from a Soft WDT call or Startup, maybe some more indirect paths!
  * Handle some strange stuff
  * Grab a copy of restart info from RTC memory.
  * Keep restart info reason REASON_EXCEPTION_RST or REASON_SOFT_WDT_RST for all others:
      * zero restart info
      * set reason to REASON_SOFT_RESTART.
      * Update RTC.
  * Call `system_restart_hook(&rst_inf)`
  * wait for uart TX fifo's to empty
  * Enter lock, do some hardware stuff
  * call `system_restart_core()` - ps.intlevel still at 3

`system_restart_core()` (IRAM) - This is the end of the line
  * Wait for SPI flashchip to idle down
  * Disable read cache.
  * pass control to reset vector

I have not done any experimentation with this one, so I cannot say much more. I am not exactly sure how to generate an event to take the path toward that `return`. Maybe something for a later investigation.

Reply:
system_restart_local is called by the system boot logic in the SDK libmain.a file.  I don't have the exact logic sequence, but it is expected to return.  The SDK will then handle the IRAM loading, clearing the BSS, copying RODATA, etc. and start the app.  Don't have a clear picture, but since a) the stack is still good enough to backtrack and b) the static globals haven't been cleared, we can intuit the above.

The way things work here is we use GCC to rename our function "system_restart_local" and then change the libmain.a's fcn to __real_... Basically, @igrr found a cool way to shim in this debug code before the stack is corrupted or the RTC clock memory (storing the exception info) is cleared.

The return you hilighted *maybe* should be   `{ __real_system_restart_local(); return; }` but, if it was a return there before it's been there a long time without any issue so I'd be loathe to open it up w/o deeper understanding.

What are you looking for, specifically?  I'm a little worried this could be pulling you away from the real problem you're trying to solve...

Reply:
I agree with @earlephilhower: maybe it shouldn't be a return, but a call to __real_system_restart_local(). That was my first thought when trying to understand the issue described.
But my second thought was exactly the same as his: if that is true, then the issue has been present for a very long time, which begs the question: why hasn't a problem been found before? Or maybe there has, but we haven't realized it?
I think it would make sense to discuss this return statement in detail elsewhere. We shouldn't hijack this PR. @mhightower83 If you're willing to pursue this, could you please open a new issue and explain the suspected problem there? Maybe link this discussion, or quote it there. That issue could then be used to track further investigation as well as experimentation.